Sources/PriorityQueueModule/Heap.swift

```diff
@@ -0,0 +1,537 @@
+
+    return (index + 1)._binaryLogarithm() & 0b1 == 0
+  }
+
+  /// Swaps the elements in the heap at the given indices.
+  @inline(__always)
+  @inlinable
+  internal mutating func _swapAt(_ i: Int, _ j: Int) {
+    let tmp = storage[i]
+    storage[i] = storage[j]
+    storage[j] = tmp
+  }
+
+  /// Returns the parent index of the given `index`
+  /// or `nil` if the index has no parent (i.e. `index == 0`).
+  @inline(__always)
+  @inlinable
+  internal func _parentIndex(of index: Int) -> Int? {
+    guard index > 0 else {
+      return nil
+    }
+
+    return (index - 1) / 2
+  }
+
+  /// Returns the grandparent index of the given `index`
+  /// or `nil` if the index has no grandparent.
+  @inline(__always)
+  @inlinable
+  internal func _grandparentIndex(of index: Int) -> Int? {
+    guard index > 2 else {
+      return nil
+    }
+
+    return (index - 3) / 4
+  }
+
+  /// Returns the first child index of the given `index`
+  /// or `nil` if the index has no children.
+  @inline(__always)
+  @inlinable
+  internal func _leftChildIndex(of index: Int) -> Int? {
+    let childIdx = index * 2 + 1
+    guard childIdx < storage.count else {
+      return nil
+    }
+
+    return childIdx
+  }
+
+  /// Returns the right child index of the given `index`
+  /// or `nil` if the index has no right child.
+  @inline(__always)
+  @inlinable
+  internal func _rightChildIndex(of index: Int) -> Int? {
+    let childIdx = index * 2 + 2
+    guard childIdx < storage.count else {
+      return nil
+    }
+
+    return childIdx
+  }
+
+  /// Returns the first grandchild index of the given `index`
+  /// or `nil` if the index has no grandchildren.
+  @inline(__always)
+  @inlinable
+  internal func _firstGrandchildIndex(of index: Int) -> Int? {
+    let grandchildIdx = index * 4 + 3
+    guard grandchildIdx < storage.count else {
+      return nil
+    }
+
+    return grandchildIdx
+  }
+
+  /// Returns the last valid grandchild index of the given `index`
+  /// or `nil` if the index has no grandchildren.
+  ///
+  /// In cases where the given index only has one grandchild, the index
+  /// returned by this function is the same as that returned by
+  /// `_firstGrandchildIndex`.
+  @inline(__always)
+  @inlinable
+  internal func _lastGrandchildIndex(of index: Int) -> Int? {
+    guard _firstGrandchildIndex(of: index) != nil else {
+      // There are no grandchildren of the node at `index`
+      return nil
+    }
+
+    return Swift.min(index * 4 + 6, storage.count - 1)
+  }
+}
+
+// MARK: -
+
+extension Heap {
+  /// Initializes a heap from a sequence.
+  ///
+  /// Utilizes [Floyd's linear-time heap construction algorithm](https://en.wikipedia.org/wiki/Heapsort#Floyd's_heap_construction).
+  ///
+  /// - Complexity: O(n), where `n` is the length of `elements`.
+  @inlinable
+  public init<S: Sequence>(_ elements: S) where S.Element == Element {
+    storage = Array(elements)
+
+    for idx in (0..<(storage.count / 2)).reversed() {
```

Comment:
Nit: put spaces around all binary operators -- `..<` and `...` shouldn't be special.

```suggestion
    for idx in (0 ..< (storage.count / 2)).reversed() {
```